src/backend/optimizer/plan/createplan.c

```diff
@@ -667,12 +667,27 @@ disuse_physical_tlist(PlannerInfo *root, Plan *plan, Path *path)
 			plan->targetlist = build_path_tlist(root, path);
 
 			/**
-			 * If plan has a flow node, ensure all entries of hashExpr
-			 * are in the targetlist.
+			 * If plan has a flow node, ensure all entries of hashExpr,
+			 * which are not of type Param, are in the targetlist.
 			 */
 			if (plan->flow && plan->flow->hashExprs)
 			{
-				plan->targetlist = add_to_flat_tlist_junk(plan->targetlist, plan->flow->hashExprs, true /* resjunk */);
```

Comment:
You've replaced `add_to_flat_tlist_junk` function with a custom logic that filter Params. But there are several places of usage `add_to_flat_tlist_junk` in the gpdb codebase that also modify target list. Why should we patch only this place?

Reply:
`add_to_flat_tlist_junk` with `hashExprs` is called in 2 more places: 
**1.** Inside `cdbpathtoplan_create_motion_plan`. This function is called from `create_motion_plan` after the call of `disuse_physical_tlist` for [subpath](https://github.com/arenadata/gpdb/blob/adb-6.x/src/backend/optimizer/plan/createplan.c#L1396-L1399) of CdbMotionPath.  The `add_to_flat_tlist_junk` is called for hashExprs in case of [redistribute motion](https://github.com/arenadata/gpdb/blob/adb-6.x/src/backend/optimizer/plan/createplan.c#L7217-L7251) . Consider example for hash join, we have outer_plan created with redistibute motion:
```
create table t2( t2c1 int, a int) distributed by (t2c1);
create table t1( t1c1 int, a int ) distributed by (t1c1);
explain verbose select * from t2 join lateral
(select * from t1 where t2.t2c1 = t1.a ) x on true;
                                           QUERY PLAN                                              
-----------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice2; segments: 3)  
   Output: t2.t2c1, t2.a, t1.t1c1, t1.a
   ->  Hash Join 
         Output: t2.t2c1, t2.a, t1.t1c1, t1.a
         Hash Cond: (t1.a = t2.t2c1)
         ->  Redistribute Motion 3:3  (slice1; segments: 3) 
               Output: t1.t1c1, t1.a
               Hash Key: t1.a
               ->  Seq Scan on public.t1 
                     Output: t1.t1c1, t1.a
         ->  Hash 
               Output: t2.t2c1, t2.a
               ->  Seq Scan on public.t2
                     Output: t2.t2c1, t2.a
 Optimizer: Postgres query optimizer
(15 rows)
```
When we are creating the outer plan for hashjoin, `create_motion_plan` is called, where for SeqScan part `disuse_physical_tlist` is called too. Then we fall into `cdbpathtoplan_create_motion_plan`  [redistribute motion](https://github.com/arenadata/gpdb/blob/adb-6.x/src/backend/optimizer/plan/createplan.c#L7217-L7251) case, where we try to add into `subplan->targetList` (subplan here corresponds to the SeqSqan on t1) `hashExprs` taken for motion locus. `hashExprs` here contain the keys on which the redistribition will perform:
```
hashExprs (
      {VAR 
      :varno 4 
      :varattno 2       <----- t1.a 
      :vartype 23 
      :vartypmod -1 
      :varcollid 0 
      :varlevelsup 0 
      :varnoold 4 
      :varoattno 2 
      :location 80
      }
   )
``` 
t1.a is being added in hashExprs in `cdbpathlocus_get_distkey_exprs` func, it iterates through equivalence class members and do several checks whether member is already in targetList ( in `cdbpullup_findEclassInTargetList`). And in this case it is already inside the targetList because of previous `disuse_physical_tlist` call, which adds t1.a to SeqScan TL, because it is in select list (if not it would be added because of WHERE clause with t1.a as an operand).

Next,  `add_to_flat_tlist_junk` will try to add to SeqScan's targetList t1.a. And again there is a check on presenece inside the TL. It presents.
Why the filtration of params is unneccesary in that case? For most of the joins, excepting nested loop, the lateral refernce will be added to all EquivalenceClasses as Var, not Param, because the lateral subquery is not really considered as a separate subquery here. Once more, I think the main problem in limit clause causing nested loop processing.  And It would be better filter out params only in that case, because I am afraid that params could occur in another cases and there would be optimization losses. But I still can't imagine and can't find another example query with param inside EquivalenceClass. In other words, I can't proof my solution strongly.
Probably Andrey Sokolov's solution is better because it prevents from adding anything into hashExprs. And all `add_to_flat_tlist_junk` will be safe in that case. 

**2.** Inside grouping_planner function the `add_to_flat_tlist_junk` is called for [window function plan's tlist.](https://github.com/arenadata/gpdb/blob/adb-6.x/src/backend/optimizer/plan/planner.c#L2581-L2589) . And in this case there is no gurantees of external Params absence. Again, probably Andrey's Approach will make it safer

Reply:
> Then we fall into cdbpathtoplan_create_motion_plan [redistribute motion](https://github.com/arenadata/gpdb/blob/adb-6.x/src/backend/optimizer/plan/createplan.c#L7217-L7251) case, where we try to add into subplan->targetList (subplan here corresponds to the SeqSqan on t1) hashExprs taken for motion locus.

only in case of `hashExprs` [contains](https://github.com/arenadata/gpdb/blob/adb-6.x/src/backend/optimizer/plan/createplan.c#L7234) a sub-plan reference. Your example plan doesn't contain subplans.